packages/utils/telemetry-utils/src/telemetryEventBatcher.ts

```diff
@@ -0,0 +1,139 @@
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+// type Data<TKey extends string> = { readonly [key in TKey]: number };
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+interface IMeasuredCodeResult<TKey extends string> {
+	telemetryProperties?: { readonly [key in TKey]: number };
+}
+
+/**
+ * @remarks The logger is expected to be used for a single event type. If the set of `telmetryProperties` is different for different events, a separate `TelemetryEventBatcher` should be created for each event type.
+ * Telemetry class that accumulates user defined telemetry metrics {@link ICustomDataMap} and sends it to the {@link  ITelemetryLoggerExt} logger provided to this class every time the {@link TelemetryEventBatcher.log} the number of calls to this function reaches a number specified by {@link TelemetryEventBatcher.threshold}.
+ * @typeparam TMetrics - The set of keys that should be logged.
+ * E.g., `keyof Foo` for logging properties `bar` and `baz` from `type Foo = { bar: number, baz: number }`.
+ *
+ * @internal
+ */
+export class TelemetryEventBatcher<TMetrics extends string> {
+	/**
+	 * `codeDuration`: stores the average duration of the code passed into the logger.
+	 * `dataSums`: stores the sum of the custom data passed into the logger.
+	 * `dataMaxes`: stores the maximum value of the custom data passed into the logger.
+	 * `dataSums` and `dataMaxes` should share identical sets of properties.
+	 */
+	private codeDuration: number = 0;
+	private dataSums: { [key in TMetrics]?: number } = {};
+	private dataMaxes: { [key in TMetrics]?: number } = {};
+
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	public constructor(
+		/**
+		 * Custom properties to include in the telemetry performance event when it is written.
+		 */
+		private readonly eventBase: ITelemetryGenericEventExt,
+
+		/**
+		 * The logger to use to write the telemetry performance event.
+		 */
+		private readonly logger: ITelemetryLoggerExt,
+
+		/**
+		 * The number of logs to accumulate before sending the data to the logger.
+		 */
+		private readonly threshold: number,
+	) {}
+
+	/**
+	 * Executes the specified code and keeps track of execution time statistics.
+	 * @param codeToMeasure - The code to be executed and measured.
+	 * @param customData - Custom data to be logged.
+	 *
+	 * @returns Whatever the passed-in code block returns.
+	 */
+	public measure<T extends IMeasuredCodeResult<TMetrics>>(
+		codeToMeasure: () => T,
+		customData: Record<TMetrics, number>,
+	): T {
+		const start = performance.now();
+		const returnValue = codeToMeasure();
+		const duration = performance.now() - start;
+
+		this.codeDuration = truncateToDecimalPlaces(
+			(this.codeDuration + duration) / this.counter,
+			6,
+		);
```

Comment:
This is not accumulating `codeDuration` correctly, since we're dividing by the counter every single time instead of first just summing all the durations and at the end dividing by the counter once. I think we want to make the variable `accumulatedDuration`, here we just keep adding to it (no need to use the rounding/truncation function, even), and when we actually send the event is when we divide by the counter, like we do for the other accumulated metrics.